### PR TITLE
Support for `Dimension(...).grain(...)` in where filter

### PR DESCRIPTION
**Pull request by DevonFulcher**
Resolves # SL-631 (internal to dbt labs)

### Description

This PR adds support for the `Dimension(...).grain(...)` syntax for the where parameter and filter spec by using the factory pattern & implementing the Query Interface protocols.

I started on this branch before Tom & I talked about the git norms in this repo. Next PR, I will be sure to make every commit meaningful.

### CODE REVIEW

**Comment by tlento** [comment]
> I started on this branch before Tom & I talked about the git norms in this repo. Next PR, I will be sure to make every commit meaningful.

Just as an aside, if you squash it's fine as long as the PR is reviewable in one chunk - I mainly care that every commit to the main source tree is focused and documented, so if squash achieves that it's totally fine.

**Review by tlento — APPROVED** [review]
Assuming I still understand this correctly I think this is good for now. Thanks for pushing these down to a more central location!

Once the current push is done we should go over all of this and see if there's an easy way to refine the interfaces further. I know Paul is doing some refactoring of the query parser itself, which I'm sure will have an effect on where we eventually land with these protocol specs.

File: metricflow/specs/where_filter_dimension.py [hunk 1]
```diff
@@ -30,39 +26,49 @@ class WhereFilterDimension(ProtocolHint[QueryInterfaceDimension]):
...
-        """Renaming the column."""
-        raise NotImplementedError
+        raise InvalidQuerySyntax("date_part isn't currently supported in the where parameter")
 
     def descending(self, _is_descending: bool) -> QueryInterfaceDimension:
         """Set the sort order for order-by."""
-        raise InvalidQuerySyntax(
-            "Can't set descending in the where clause. Try setting descending in the order_by clause instead"
-        )
+        raise InvalidQuerySyntax("descending is invalid in the where parameter")
```
[the remark this answers is not in the document]
Comment (DevonFulcher) [reply metricflow/specs/where_filter_dimension.py:56]: I think the spec is fine. We don't have an LSP for the Jinja syntax, so I think providing a more specific error message will help our users more than a generic type error.

File: metricflow/specs/where_filter_dimension.py [hunk 2]
```diff
@@ -30,39 +26,49 @@ class WhereFilterDimension(ProtocolHint[QueryInterfaceDimension]):
...
+        return self._column_association_resolver.resolve_spec(
+            self.time_dimension_spec or self.dimension_spec
+        ).column_name
 
 
 class WhereFilterDimensionFactory(ProtocolHint[QueryInterfaceDimensionFactory]):
     """Creates a WhereFilterDimension.
 
-    Each call to `create` adds a DimensionSpec to dimension_specs.
+    Each call to `create` adds a DimensionSpec to created.
```
[the remark this answers is not in the document]
Comment (DevonFulcher) [reply metricflow/specs/where_filter_dimension.py:71]: Thank you!

File: metricflow/specs/dimension_spec_resolver.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            time_granularity=time_granularity_name,
+            entity_path=(
+                tuple(EntityReference(element_name=arg) for arg in entity_path) + structured_name.entity_links
+            ),
+        )
+        assert call_parameter_set in self._call_parameter_sets.time_dimension_call_parameter_sets
+        return TimeDimensionSpec(
+            element_name=call_parameter_set.time_dimension_reference.element_name,
+            entity_links=call_parameter_set.entity_path,
+            time_granularity=call_parameter_set.time_granularity,
```
[the remark this answers is not in the document]
Comment (DevonFulcher) [reply metricflow/specs/dimension_spec_resolver.py:53]: I'm not sure! This code was moved here from elsewhere in this PR, which was also moved there from somewhere that @plypaul put it. So, Paul is the original author, and I just moved it here. He may know if this is duplicative of somewhere else in MF.

File: metricflow/specs/where_filter_dimension.py [hunk 4]
```diff
@@ -30,39 +26,49 @@ class WhereFilterDimension(ProtocolHint[QueryInterfaceDimension]):
...
+        self.time_dimension_spec: Optional[TimeDimensionSpec] = None
 
-    def grain(self, _grain: str) -> QueryInterfaceDimension:
+    def grain(self, time_granularity_name: str) -> QueryInterfaceDimension:
         """The time granularity."""
-        raise NotImplementedError
+        self.time_dimension_spec = self._dimension_spec_resolver.resolve_time_dimension_spec(
+            self._name, TimeGranularity(time_granularity_name), self._entity_path
+        )
+        return self
```
[the remark this answers is not in the document]
Comment (DevonFulcher) [reply metricflow/specs/where_filter_dimension.py:48]: Can you clarify what you mean by `this` and `these` 😃?

File: metricflow/specs/where_filter_dimension.py [hunk 5]
```diff
@@ -30,39 +26,49 @@ class WhereFilterDimension(ProtocolHint[QueryInterfaceDimension]):
...
+        entity_path: Sequence[str],
+        call_parameter_sets: FilterCallParameterSets,
+        column_association_resolver: ColumnAssociationResolver,
+    ) -> None:
+        self._dimension_spec_resolver = DimensionSpecResolver(call_parameter_sets)
+        self._column_association_resolver = column_association_resolver
+        self._name = name
+        self._entity_path = entity_path
+        self.dimension_spec = self._dimension_spec_resolver.resolve_dimension_spec(name, entity_path)
+        self.time_dimension_spec: Optional[TimeDimensionSpec] = None
```
[the remark this answers is not in the document]
Comment (DevonFulcher) [reply metricflow/specs/where_filter_dimension.py:41]: > Should we be using the type system to enforce that exactly one of these is set?

Do you mean something like:
```
self.spec: Union[DimensionSpec, TimeDimensionSpec] = self._dimension_spec_resolver.resolve_dimension_spec(name, entity_path)
```
I had started with that, but I think that would require the use of `isinstance` to disambiguate between the two types later when they are added to the lists. Paul steered me away from the use of `isinstance`.


> protection against someone using a time dimension as a non-time dimension

Do you mean specifying grain on a non-time dimension? Or not specifying grain on a time dimension? I had thought that is validated after the specs are returned from `create_from_where_filter`. Is that not the case? I agree that should be checked somewhere

**Pull request merged**